### DISCUSSION TITLE
[Routing] Incorrect routing on custom REST API methods (with demo-repo)

### DISCUSSION THREAD

**Discussion by sebastianstucke87** [symfony/symfony#50135] [Q&A] [answered]
### Symfony version(s) affected

6.2

### Description

I have 2 routes:
- standard: `/api/example/{id}`
- custom: `/api/example/{id}:some-action`

When I execute the standard-route, I get this correct response:
```JSON
{
  "php-method": "standard",
  "ID": "abc123"
}
```

But when I execute the custom-route, I get this **incorrect** response:
```JSON
{
  "php-method": "standard", // should be "custom"
  "ID": "abc123:some-action" // should be "abc123"
}
```

### How to reproduce

Please see this **demo-repo**: https://github.com/sebastianstucke87/symfony-routing-bug-with-custom-methods

(It is a fresh symfony projection with the Controller, the rest-client .http-files a detailed README.md 👍)

In summary, this is the Controller:
```php
final class ExampleController
{
    #[Route(path: '/api/example/{id}', methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'])]
    public function standard(string $id): Response
    {
        return new JsonResponse(
            [
                'php-method' => __METHOD__,
                'ID' => $id,
            ]
        );
    }

    #[Route(path: '/api/example/{id}:some-action', methods: ['GET', 'POST'])]
    public function custom(string $id): Response
    {
        return new JsonResponse(
            [
                'php-method' => __METHOD__,
                'ID' => $id,
            ]
        );
    }
}

```

### Additional Context

REST API Custom Methods:
- https://cloud.google.com/apis/design/custom_methods
    - https://cloud.google.com/apis/design/custom_methods#common_custom_methods
- https://google.aip.dev/136

**Comment by wouterj** [comment, marked as answer]
You need to define route parameters requirements to avoid everything being eaten by the `{id}` in the first route: https://symfony.com/doc/current/routing.html#parameters-validation

**Reply by sebastianstucke87** [reply]
Ah thanks!  

So in my example this would result in:

```php
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\Routing\Requirement\Requirement;

final class ExampleController
{
    #[Route(path: '/api/example/{id}', requirements: ['id' => Requirement::UUID], methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'])]
    public function standard(string $id): Response
    {
        return new JsonResponse(
            [
                'php-method' => __METHOD__,
                'ID' => $id,
            ]
        );
    }

    #[Route(path: '/api/example/{id}:some-action', requirements: ['id' => Requirement::UUID],  methods: ['GET', 'POST'])]
    public function custom(string $id): Response
    {
        return new JsonResponse(
            [
                'php-method' => __METHOD__,
                'ID' => $id,
            ]
        );
    }
}
```

```http
### Standard API
GET https://127.0.0.1:8000/api/example/90c3212e-d70a-4e19-b265-8a272d8c123e
Accept: *

### Custom API
GET https://127.0.0.1:8000/api/example/90c3212e-d70a-4e19-b265-8a272d8c123e:some-action
Accept: *
```

**Reply by sebastianstucke87** [reply]
Why doesn't it work by default, though? Just curious, because it seems like a bit of overhead to do this for every potential param/route..

**Reply by wouterj** [reply]
We can't make any assumption on the format of a route parameter. E.g. some people use numeric IDs, some people UUIDs, maybe people also use a syntax of `uuid:1203-...`. And a route parameter doesn't even have to be an ID, so the format is even lesser known.

So the Routing component doesn't have any default regex to match patterns to, except that it by default does not allow route parameters to contain a `/`.
